lib/collection/src/shards/queue_proxy_shard.rs

```diff
@@ -62,6 +65,43 @@ impl QueueProxyShard {
         }
     }
 
+    /// Queue proxy the given local shard and point to the remote shard, from a specific WAL version.
+    ///
+    /// This queues all (existing) updates from a specific WAL `version` and onwards. In other
+    /// words, this will ensure we transfer updates we already have and all new updates from a
+    /// specific point in our WAL. The `version` may be in the past, but must always be within
+    /// range of the current WAL.
+    ///
+    /// # Errors
+    ///
+    /// This fails if the given `version` is not in bounds of our current WAL. If the given
+    /// `version` is to old or to new, queue proxy creation is rejected.
+    pub fn new_from_version(
+        wrapped_shard: LocalShard,
+        remote_shard: RemoteShard,
+        wal_keep_from: Arc<AtomicU64>,
+        version: u64,
+    ) -> Result<Self, (LocalShard, CollectionError)> {
+        // Lock WAL until we've successfully created the queue proxy shard
+        let wal = wrapped_shard.wal.clone();
+        let wal_lock = wal.lock();
+
+        // If start version is not in current WAL bounds, we cannot reliably transfer WAL
+        let (first_idx, last_idx) = (wal_lock.first_physical_index(), wal_lock.last_index());
+        if version < first_idx || version > last_idx {
+            return Err((wrapped_shard, CollectionError::service_error(format!("Cannot create queue proxy shard from version {version} because it is out of WAL bounds ({first_idx}..={last_idx})"))));
```

Comment:
`[{first_idx}, {last_ids}]`? Like, math notation for inclusive range?
(Just a suggestion, though. I'm cool with either.)

Reply:
Good idea, I'll do that to make it more explicit.

I'll also do `(first_idx..=last_idx).contains()` to more closely match the condition.

Reply:
It'll be `(first_idx, last_idx)` by the way, as it is inclusive.

Reply:
`[start, end]` is inclusive. `(x, y)` is exclusive. But I think we may better leave it as `..=` then? :D
https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints

Reply:
:man_facepalming:

Reply:
Fixed.